qiskit/circuit/quantumcircuit.py

```diff
@@ -1837,31 +1844,43 @@ def parameters(self):
 
         # parameters in global phase
         if isinstance(self.global_phase, ParameterExpression):
-            return params.union(self.global_phase.parameters)
+            params = params.union(self.global_phase.parameters)
+
+        # sort the parameters
+        sorted_by_name = sorted(list(params), key=functools.cmp_to_key(_compare_parameters))
```

Comment:
i'm computing this in my head. But this seems to be more straightforward:

```suggestion
        sorted_by_name = sorted(list(params), key=lambda  param: param.name)
```

Reply:
This doesn't do the correct thing unfortunately because we need to respect the order of the parameter vector. Sorting by `param.name` will sort the following set
```
{a, z, x[0], x[1], x[2], x[3], x[10], x[11]} 
```
as
```
[a, x[0], x[1], x[10], x[11], x[2], x[3], z]
```
since, alphabetically the string `10` is smaller than `2` as it begins with a `1`. But we want 
```
[a, x[0], x[1], x[2], x[3], x[10], x[11], z]
```

Reply:
Maybe a test for that case?

Reply:
Good point, added one 🙂